File: 프로그래머스/3/42898. 등굣길/등굣길.py
# 가로 m x 세로 n 크기 맵
# 물에 잠긴 지역 좌표 2차원 배열: puddles(1 based)
# 오른쪽, 아래쪽으로만 움직여 최단거리 개수 -> 1,000,000,007으로 나눈 나머지 return
# bfs -> 모든 경로 탐색 하므로 비효율적
# 최단거리 경로의 "개수" 이므로 dp로 해결!
# 경로 개수를 dp 테이블에 저장해 점진적으로 값 계산
# 점화식
# dp[i][j] = dp[i-1][j]+dp[i][j-1] # 웅덩이 아님
# dp[i][j]= 0 # 웅덩이 

def solution(m, n, puddles):
    answer = 0
    MOD = 1_000_000_007
    
    dp = [[0]*(m+1) for _ in range(n+1)] # 1based
    # 시작점 초기화
    dp[1][1] = 1
    # 웅덩이 설정
    for x, y in puddles:
        dp[y][x] = -1 # y,x 순서 주의
        
    # dp 계산
    for i in range(1, n+1):
        for j in range(1, m+1):
            if dp[i][j]== -1: # 웅덩이
                dp[i][j]=0
            elif i==1 and j==1:
                continue
            else:
                dp[i][j]=(dp[i-1][j]+dp[i][j-1])%MOD

    return dp[n][m]

File: 프로그래머스/3/42898. 등굣길/등굣길.java
import java.util.*;

class Solution {
    public int solution(int m, int n, int[][] puddles) {
        int answer = 0;
        // overflow 방지
        int MOD= 1000000007;
        
        // 1. 테이블 정의: 1 based
        int[][] dp = new int[n+1][m+1];
        
        // 3. 초기값 설정
        dp[1][1]=1;

        // System.out.println(Arrays.deepToString(dp));
        
        // puddles를 조건식으로 처리하기 위해 boolean값으로 바꾼다.
        boolean[][] isPu = new boolean[n+1][m+1];
        
        for(int[] puddle: puddles){
            isPu[puddle[1]][puddle[0]]=true; // puddles는 x,y 형식이므로 순서주의
        }
            
        // 2. 점화식 도출
        for(int i=1; i<=n; i++){
            for(int j=1; j<=m; j++){
                if(isPu[i][j]){
                    dp[i][j]=0; // 물 웅덩이라면 경로 0
                }
                else{
                    // 오른쪽으로 간 경우(=왼쪽에서 온 경우)
                    dp[i][j]=(dp[i][j]+ dp[i-1][j])%MOD;
                    // 아래로 간 경우(=위쪽에서 온 경우)
                    dp[i][j]=(dp[i][j]+ dp[i][j-1])%MOD;
                        
                }
            }
        }
        
        return dp[n][m];
    }
}